### ISSUE TITLE
Is there a way mlContext.Transform can be debugged ?

### ISSUE THREAD

**Issue by ketankalia** [dotnet/machinelearning-samples#332]
HI,
I am trying to convert the ImageClassification sample into something which can be used on MNIST. I know a mathematical model for MNIST already exists in samples, but I like to experiment in order to understand better :)
I am using v 0.10 TensorFlow sample as the base for this. For the model I am using a pre-trained MNIST model which gives good results when used  through Keras and Tensorflow (python). But same model if used in Tensorflow in  ML.Net gives incorrect results.

My gut feeling is that this code  which transforms image of a number into what the MNIST model expects (grayscale, scaled to 255) is incorrect. But, I am not sure how to debug or view the transformed image. Can you please advise what I can do to debug the transformed image.

`var pipeline = mlContext.Transforms.LoadImages(imageFolder: imagesFolder, columns: (outputColumnName: ImageReal, inputColumnName: nameof(ImageNetData.ImagePath)))
            .Append(mlContext.Transforms.Resize(outputColumnName: ImageReal, imageWidth: ImageNetSettings.imageWidth, imageHeight: ImageNetSettings.imageHeight, inputColumnName: ImageReal))
            .Append(mlContext.Transforms.ConvertToGrayscale((outputColumnName: ImageReal, inputColumnName: ImageReal)))
            .Append(mlContext.Transforms.ExtractPixels(columns: new[] { new ImagePixelExtractorTransformer.ColumnInfo(name: InceptionSettings.inputTensorName, colors: ImagePixelExtractorTransformer.ColorBits.Alpha, inputColumnName: ImageReal, interleave: ImageNetSettings.channelsLast, scale=255, asFloat: true) }))
            .Append(mlContext.Transforms.ScoreTensorFlowModel(modelLocation: modelLocation, outputColumnNames: new[] { InceptionSettings.outputTensorName }, inputColumnNames: new[] { InceptionSettings.inputTensorName }));
`

Thank you,
Ketan

**Comment by CESARDELATORRE** [comment]
Adding @jignparm - Jignesh, Can you help on this question?

**Comment by yaeldekel** [comment]
Hi @ketankalia, thank you for your question. I think that you are right, it is probably some discrepancy between the pixel format the model expects and the format it actually sees (I suspect it is a column-major/row-major issue). Is there a way that you could share the model and a couple of the images you were using so that I can look into it?

**Comment by ketankalia** [comment]
Hi @yaeldekel/ @CESARDELATORRE , 
First of all thank you for replying.
Please find attached the model (.pb file) in zipped file and 2 images which are in 28X28 format.
I used the preview option to look at the data transformations and saw that all values are equal to 255. 
![2](https://user-images.githubusercontent.com/13715128/55094264-10d7e400-5084-11e9-8b26-2349316156ee.jpg)
![0](https://user-images.githubusercontent.com/13715128/55094265-11707a80-5084-11e9-96bb-474481ba35a2.jpg)
[classify_mnist_graph_def.zip](https://github.com/dotnet/machinelearning-samples/files/3014292/classify_mnist_graph_def.zip)

This is the code I am using to split up data transformation pipeline and get a preview of data. 
Only the last line provides a preview, rest of the preview statements do not seem to provide any useful information. The preview in last statement provides all pixel values as 255 as if it is a complete black image, but as you can see above that is not the case. I am not able to understand why that is happening, maybe it has something to do with the transformations  I am doing. Please advise. 

```
var pipeline = mlContext.Transforms.LoadImages(imageFolder: imagesFolder, columns: (outputColumnName: ImageReal, inputColumnName: nameof(ImageNetData.ImagePath)));
 var transData1 = pipeline.Preview(data);
 var dopipe1 = pipeline.Append(mlContext.Transforms.Resize(outputColumnName: ImageReal, imageWidth: ImageNetSettings.imageWidth, imageHeight: ImageNetSettings.imageHeight, inputColumnName: ImageReal));
 var transData2 = dopipe1.Preview(data);
 // var dodoh = dodo1.Append(mlContext.Transforms.ConvertToGrayscale((outputColumnName: ImageReal, inputColumnName: ImageReal)));
 //  var dodo2 = dodo1.Append(mlContext.Transforms.ExtractPixels(columns: new[] { new ImagePixelExtractorTransformer.ColumnInfo(name: InceptionSettings.inputTensorName, inputColumnName: ImageReal, interleave: ImageNetSettings.channelsLast, offset: ImageNetSettings.mean) }));
//            var dodo2 = dodo1.Append(mlContext.Transforms.ExtractPixels(columns: new[] { new ImagePixelExtractorTransformer.ColumnInfo(name: InceptionSettings.inputTensorName, colors: ImagePixelExtractorTransformer.ColorBits.Alpha, inputColumnName: ImageReal, interleave: ImageNetSettings.channelsLast,scale:1 ,asFloat: true) }));
var dopipe2 = dopipe1.Append(mlContext.Transforms.ExtractPixels(columns: new[] { new ImagePixelExtractorTransformer.ColumnInfo(name: InceptionSettings.inputTensorName, colors: ImagePixelExtractorTransformer.ColorBits.Alpha, inputColumnName: ImageReal, interleave: ImageNetSettings.channelsLast) }));
var dopipe3 = dopipe2.Append(mlContext.Transforms.ScoreTensorFlowModel(modelLocation: modelLocation, outputColumnNames: new[] { InceptionSettings.outputTensorName }, inputColumnNames: new[] { InceptionSettings.inputTensorName }));

var transData = dopipe3.Preview(data);

```

**Comment by yaeldekel** [comment]
Hi @ketankalia .
Thank you for the images and model. I started looking into your issue, and while I have found some changes that you should make to your code, I am still trying to figure out how to make the predictions better. The changes that you should make are:

- Uncomment the grayscale transformation. This ensures that the R,G and B components all have the same value.
- Change the ExtractPixel transformation to extract `ImagePixelExtractorTransformer.ColorBits.Red` instead of `Alpha`.

The things that I haven't figure out yet is whether you need to add an offset/scale when you extract the pixels, and whether the images need to be rotated. Could you send me the featurization code that you used in Python to score the images with TensorFlow?

**Comment by yaeldekel** [comment]
I also want to verify how you classify the images using this model, are you taking the index of the maximal value from the "logits" layer?

**Comment by ketankalia** [comment]
Yes @yaeldekel

**Comment by yaeldekel** [comment]
Hi @ketankalia , can you post the Python featurization code that you used so that I can compare it to the ML.NET featurization?

**Comment by prathyusha12345** [comment]
Closing the issue. Please reopen if you encounter any problem, OK? Thanks 👍

**Comment by ketankalia** [comment]
> Hi @ketankalia , can you post the Python featurization code that you used so that I can compare it to the ML.NET featurization?


Hi @yaeldekel,
Sorry for this super-late reply. I have been busy some place else. 
I think this is the code you had asked for. let me know otherwise.
'''
            var dataProcessPipelineMemoryDirect = mlContext.Transforms.Conversion.MapValueToKey(outputColumnName: LabelAsKey, inputColumnName: "Label")
                           .Append(mlContext.Transforms.ResizeImages(outputColumnName: "image_object_resized",
                                                                     imageWidth: ImageSettingsForTFModel.imageWidth, imageHeight: ImageSettingsForTFModel.imageHeight,
                                                                     inputColumnName: nameof(ImageInputData.Image)))
                           .Append(mlContext.Transforms.ConvertToGrayscale(outputColumnName: "image_object_resized1", inputColumnName: "image_object_resized"))
                           .Append(mlContext.Transforms.ExtractPixels(outputColumnName: "linku_input", inputColumnName: "image_object_resized1",
                                                                      interleavePixelColors: ImageSettingsForTFModel.channelsLast,
                                                                      //                                            offsetImage: ImageSettingsForTFModel.mean,
                                                                      colorsToExtract: Microsoft.ML.Transforms.Image.ImagePixelExtractingEstimator.ColorBits.Blue, outputAsFloatArray: true,
                                                                      scaleImage: ImageSettingsForTFModel.scale))  //for Inception v3 needs scaleImage: set to 1/255f. Not needed for InceptionV1. 
                           .Append(tfModel.ScoreTensorFlowModel(outputColumnNames: new[] { "main_output/BiasAdd" },
                                     inputColumnNames: new[] { "linku_input" },
                                     addBatchDimensionInput: false));
'''
Thanks,
Ketan
